File: console/lib/credentials.ts
// lib/credentials.ts — shared types + pure derivations for the Credentials screen (story 8.6).
//
// The screen's data contract mirrors the apiserver's GET /api/credentials projection
// (internal/apiserver/credentials.go). All presentation derivations live here as PURE functions so
// the component stays a thin render and the derivations are unit-testable (Vitest) without a DOM.
//
// Honesty rules carried from the spec (8.6 / 8.8b no-placeholder discipline / FR-I3):
//   - an expiry the read model cannot know renders "—" (unknown), NEVER a fabricated horizon;
//   - health is the closed set connected | refreshing | expired | unknown — everything else maps
//     to unknown rather than guessing.

/** One row of the credentials table (mirror of the apiserver's AgentCredentialRow). */
export interface AgentCredentialRow {
  agent: string;
  namespace: string;
  runtime: string;
  model?: string;
  /** "namespace/name" of the per-user Secret (FR-G1 — never a master credential). */
  credentialRef: string;
  credentialClass?: string;
  expiresAt?: string;
  expiresKnown: boolean;
  health: string;
  pausedRuns?: PausedRunRef[];
}

/** One Run paused on a credential hold (deep-links to the run stream 8.2 / detail 8.11). */
export interface PausedRunRef {
  name: string;
  reason: string;
  since?: string;
}

/** GET /api/credentials payload (mirror of the apiserver's CredentialsOverview). */
export interface CredentialsOverview {
  team: string;
  agents: AgentCredentialRow[];
  connectClaude: boolean;
}

/** Classified fetch outcome for the screen — 401/403/404 collapse (existence-hiding, AC4-style). */
export type CredentialsOutcome =
  | { kind: "ok"; data: CredentialsOverview }
  | { kind: "not-found" } // unauthenticated / denied / no team — indistinguishable by design
  | { kind: "unconfigured" } // apiserver's documented 501 (read model not wired yet)
  | { kind: "error" }; // 5xx / network — the apiserver itself is unhappy

/** Token-type label for the row: the credential-class annotation when present, else the shape of
 *  the Secret ref is unknown to the console — render honestly rather than guess OAuth-vs-API-key. */
export function tokenTypeLabel(row: AgentCredentialRow): string {
  switch (row.credentialClass) {
    case "claude_oauth":
      return "OAuth";
    case "api_key":
      return "API key";
    case "byo_endpoint":
      return "BYO endpoint";
    case "human-seat":
      return "OAuth · seat";
    case "service-account":
      return "Service account";
    default:
      return "—";
  }
}

/** Expiry cell text: known horizon formatted compactly; unknown renders the em dash (never a
 *  fabricated number). Static keys ("API key") surface as "— (static)" only when the class says so. */
export function expiryLabel(row: AgentCredentialRow, now: Date = new Date()): string {
  if (row.credentialClass === "api_key" || row.credentialClass === "byo_endpoint") {
    return "— (static)";
  }
  if (!row.expiresKnown || !row.expiresAt) {
    return "—";
  }
  const at = new Date(row.expiresAt);
  const diffMs = at.getTime() - now.getTime();
  if (diffMs <= 0) {
    return `expired ${formatAgo(-diffMs, now, at)} ago`;
  }
  return `in ${formatDuration(diffMs)}`;
}

/** True when a known horizon is within the soon-window (soonThresholdMs) — drives the
 *  "Expiring soon" badge. Unknown horizons are never "soon" (honesty over alarm). */
export function expiringSoon(row: AgentCredentialRow, now: Date = new Date(), soonThresholdMs = 45 * 60 * 1000): boolean {
  if (!row.expiresKnown || !row.expiresAt) return false;
  const diff = new Date(row.expiresAt).getTime() - now.getTime();
  return diff > 0 && diff <= soonThresholdMs;
}

/** Health badge state for a row: the closed set + the paused/soon overlays the mock 05 shows. */
export type HealthBadge = {
  label: string;
  tone: "ok" | "warn" | "bad" | "idle";
};

export function healthBadge(row: AgentCredentialRow, now: Date = new Date()): HealthBadge {
  if (row.health === "expired" || row.pausedRuns?.length) {
    return { label: "Expired · paused", tone: "bad" };
  }
  if (row.health === "refreshing") {
    return { label: "Refreshing", tone: "warn" };
  }
  if (row.health === "connected") {
    if (expiringSoon(row, now)) {
      return { label: "Expiring soon", tone: "warn" };
    }
    return { label: "Valid", tone: "ok" };
  }
  return { label: "Unknown", tone: "idle" };
}

/** The paused-on-expiry banner selection (8.6 AC / S10): the most recent credential hold across
 *  the rows — the screen's clearest operator signal. Null when nothing is held. Timestamps are
 *  compared as parsed instants, never as strings: Go's encoding/json emits RFC3339Nano, where a
 *  sub-second suffix would sort before "Z" lexicographically and pick the WRONG run (PR #87
 *  review). NaN comparisons are false, preserving first-wins when a `since` is absent. */
export interface BannerHold {
  agent: string;
  run: PausedRunRef;
}

const holdTs = (r: PausedRunRef): number => (r.since ? Date.parse(r.since) : Number.NaN);

export function bannerHold(rows: AgentCredentialRow[]): BannerHold | null {
  let worst: BannerHold | null = null;
  for (const row of rows) {
    for (const run of row.pausedRuns ?? []) {
      if (worst === null || holdTs(run) > holdTs(worst.run)) {
        worst = { agent: row.agent, run };
      }
    }
  }
  return worst;
}

/** Relative "x ago" for a past instant (banner copy mirrors mock 05's "expired 6m ago"). */
function formatAgo(diffMs: number, _now: Date, at: Date): string {
  void _now;
  void at;
  return formatDuration(diffMs);
}

/** Compact human duration: 41m / 5h 12m / 2h 48m / 9d — mock 05's expiry idiom. */
export function formatDuration(ms: number): string {
  const m = Math.round(ms / 60000);
  if (m < 1) return "<1m";
  if (m < 60) return `${m}m`;
  const h = Math.floor(m / 60);
  const rm = m % 60;
  if (h < 24) return rm ? `${h}h ${rm}m` : `${h}h`;
  const d = Math.floor(h / 24);
  return `${d}d`;
}

/** Classify a fetch status into the screen's outcome vocabulary. */
export function classifyCredentialsStatus(status: number): Exclude<CredentialsOutcome, { kind: "ok" }>["kind"] {
  if (status >= 200 && status < 300) throw new Error("ok is not a failure kind");
  if (status === 401 || status === 403 || status === 404) return "not-found";
  if (status === 501) return "unconfigured";
  return "error";
}

// ── BYO service-account create form (ISI-3983) ────────────────────────────────
//
// The v1 credential WRITE path is a bring-your-own service-account key paste (POST
// /api/credentials, ISI-3679/ISI-3937). The (runtime, class) options mirror the credinject
// injection table (pkg/credinject): claude-code / openclaw / hermes carry a service-account
// key. The human-seat OAuth class is NOT paste-able — it is minted by the Connect Claude flow
// (ISI-2899) and the apiserver answers a documented 501 for it, so it is deliberately absent
// from the paste form (the button above stays the honest, coming-soon seam for it).

/** A runtime offered by the paste form (mirror of api.RuntimeType* + the credinject table). */
export interface CreateRuntimeOption {
  value: string;
  label: string;
}

/** The service-account runtimes the injection table maps for a pasted key. */
export const CREATE_RUNTIME_OPTIONS: CreateRuntimeOption[] = [
  { value: "claude-code", label: "Claude Code" },
  { value: "openclaw", label: "OpenClaw" },
  { value: "hermes", label: "Hermes" },
];

/** The only class the paste form writes: a long-lived service-account key (never human-seat). */
export const CREATE_CLASS = "service-account";

/** The write body for POST /api/credentials. `value` is write-only (never serialised back out). */
export interface CredentialCreateInput {
  name: string;
  runtime: string;
  value: string;
  /** Fleet-admin routing hint (ISI-3937). Omitted for a bound single-team caller. */
  teamId?: string;
}

/**
 * Build the POST /api/credentials JSON body from the form input. Pure + unit-testable: trims the
 * name (a DNS-1123 Secret name), pins the service-account class, forwards the value verbatim, and
 * includes `teamId` ONLY when a fleet admin has picked a team (an empty hint is inert upstream).
 */
export function credentialCreateBody(input: CredentialCreateInput): Record<string, string> {
  const body: Record<string, string> = {
    name: input.name.trim(),
    runtime: input.runtime,
    class: CREATE_CLASS,
    value: input.value,
  };
  if (input.teamId) body.teamId = input.teamId;
  return body;
}

/** Outcome of a create POST, mapped to honest screen copy (status surfaced verbatim by the BFF). */
export type CreateOutcome =
  | "created" // 201 — Secret written, secretRef returned (never the value)
  | "select-team" // 400 — fleet admin must pick a team (ISI-3937 semantics)
  | "invalid" // 422 — field validation (name/runtime/value)
  | "conflict" // 409 — a credential with this name already exists
  | "denied" // 401/403/404 — deny/not-found collapse (existence-hiding)
  | "unsupported" // 501 — human-seat, provisioned by Connect Claude not a paste
  | "error"; // 5xx / network — the store itself is unhappy

export function classifyCreateStatus(status: number): CreateOutcome {
  if (status >= 200 && status < 300) return "created";
  if (status === 400) return "select-team";
  if (status === 409) return "conflict";
  if (status === 422) return "invalid";
  if (status === 501) return "unsupported";
  if (status === 401 || status === 403 || status === 404) return "denied";
  return "error";
}

/** One squad row for the fleet-admin team picker (mirror of the fleetlist TeamListEntry). */
export interface CredentialTeamOption {
  name: string;
  namespace: string;
  uid: string;
}

/** GET /api/squad/teams payload (subset the picker needs). `teams` may arrive null/absent. */
export interface CredentialTeamList {
  teams?: CredentialTeamOption[] | null;
  fleet?: boolean;
}

File: internal/apiserver/credentials.go
package apiserver

import (
	"context"
	"errors"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"sigs.k8s.io/controller-runtime/pkg/client"

	ksquadv1 "github.com/K8squad/K8squad/api/v1alpha1"
	"github.com/K8squad/K8squad/internal/discussion"
)

// ============================================================================
// Credential / auth-state read model (story 8.6 / ISI-2902) — the projection
// GET /api/credentials answers for the console's Credentials screen.
// ============================================================================
//
// Source of truth is the controller-runtime informer cache (same discipline as overview.go):
// Agents and Runs are CRDs, so the per-agent credential surface (BYO Secret ref, runtime, and any
// Paused-on-credential Runs) is the cache's level-triggered projection of etcd. The projection is
// Team-SCOPED (AuthorContext.TeamID → Team namespace, §12.1 tenancy root) — a caller never sees
// another squad's credential rows.
//
// Health derivation (8.6 AC + CEO 2026-08-12): per-agent health is *connected / refreshing /
// expired*, derived from the Run conditions the reconciler writes (run_types.go documents
// Paused(auth_failure) / Paused(rate_limited); Epic 7.4 adds the credential reason family). The
// screen must show a CLEAR paused-on-expiry signal — so a Paused Run whose condition reason names
// a credential failure marks that Agent's row expired·paused and feeds the banner. What the cache
// CANNOT see (token expiry horizons, refresh progress — the 7.7 credential controller's Secret
// writes, which are also where *connected* and *refreshing* will be derived from once ISI-2899
// lands) is surfaced as **unknown** rather than fabricated: absence of a paused Run is NOT
// evidence a credential works, so the projection never seeds a green badge on zero knowledge
// (no-placeholder discipline, 8.8b / FR-I3).
//
// FR-G1 footer fact is structural: every row is a per-user Secret ref read off the Agent spec;
// there is no master-credential field anywhere to project.

// Health states for one Agent's credential (8.6 CEO 2026-08-12: connected / refreshing / expired).
const (
	CredHealthConnected  = "connected"
	CredHealthRefreshing = "refreshing"
	CredHealthExpired    = "expired"
	CredHealthUnknown    = "unknown"
)

// CredentialCondition reasons (metav1.Condition.Reason) that mark a credential hold. The
// reconciler's family per arch §8/§10: auth failures and credential expiry pause the Run; the
// exact vocabulary is pinned by Epic 7.4/ISI-2898 (credential_expired / credential_rotated /
// endpoint_unreachable) and K8s convention may PascalCase it (CredentialExpired) — matched on a
// case/separator-normalized token so a casing change upstream needs no console change.
var credentialHoldReasons = map[string]bool{
	"authfailure":         true,
	"credexpired":         true,
	"credentialexpired":   true,
	"credentialrotated":   true,
	"endpointunreachable": true,
}

// isCredentialHold matches the reason against the hold vocabulary on a normalization of it —
// lowercased, separator-stripped — so a casing or separator change upstream needs no console
// change. Runes outside ASCII alphanumerics are DROPPED, never narrowed: rune is int32 and
// byte(r) is a lossy G115 overflow conversion, while k8s condition reasons are ASCII-validated
// so wide runes can only be noise around the vocabulary, never part of it.
func isCredentialHold(reason string) bool {
	var b strings.Builder
	for _, r := range strings.ToLower(strings.TrimSpace(reason)) {
		switch r {
		case '_', '-', ' ', '.':
			continue
		}
		if r >= '0' && r <= '9' || r >= 'a' && r <= 'z' {
			b.WriteRune(r)
		}
	}
	return credentialHoldReasons[b.String()]
}

// PausedRunRef is one Run paused on a credential hold, for the row's RUNS cell and the banner's
// deep link (run detail 8.11 / stream 8.2).
type PausedRunRef struct {
	Name   string     `json:"name"`
	Reason string     `json:"reason"`
	Since  *time.Time `json:"since,omitempty"`
}

// AgentCredentialRow is one row of the credentials table (mock 05): agent identity, runtime,
// credential Secret ref, token class, expiry horizon, health, and any paused Runs.
type AgentCredentialRow struct {
	Agent           string         `json:"agent"`
	Namespace       string         `json:"namespace"`
	Runtime         string         `json:"runtime"`
	Model           string         `json:"model,omitempty"`
	CredentialRef   string         `json:"credentialRef"` // "namespace/name" of the per-user Secret
	CredentialClass string         `json:"credentialClass,omitempty"`
	ExpiresAt       *time.Time     `json:"expiresAt,omitempty"` // nil ⇒ unknown (no controller data)
	ExpiresKnown    bool           `json:"expiresKnown"`
	Health          string         `json:"health"` // connected|refreshing|expired|unknown
	PausedRuns      []PausedRunRef `json:"pausedRuns,omitempty"`
}

// CredentialsOverview is the GET /api/credentials payload: the Team-scoped agent rows plus the
// Connect-Claude surface state (7.7 one-click OAuth — its backing lands with ISI-2899; until then
// the field is honestly false and the button renders its not-configured state).
type CredentialsOverview struct {
	Team          string               `json:"team"`
	Agents        []AgentCredentialRow `json:"agents"`
	ConnectClaude bool                 `json:"connectClaude"`
}

// CredentialOverviewReader projects the per-agent credential surface for a single Team (by K8s
// object UID — the AuthorContext.TeamID). Production wires the cache-backed reader; tests wire a
// fake client.Reader. A reader MUST scope strictly to teamUID.
type CredentialOverviewReader interface {
	Credentials(ctx context.Context, teamUID string) (CredentialsOverview, error)
}

// ClientCredentialReader is the production CredentialOverviewReader over any client.Reader (the
// informer cache in the host; a fake client in tests). Read-only. Team UIDs are immutable for an
// object's lifetime, so the UID→namespace resolution is memoized after its first list rather
// than re-scanning every Team on every request.
type ClientCredentialReader struct {
	reader client.Reader
	// mu guards teamNS: one reader is shared across HTTP request goroutines
	// (cmd/apiserver/main.go), so the memo must be race-safe (PR #87 re-review).
	mu     sync.RWMutex
	teamNS map[string]string // teamUID → namespace (immutable once resolved)
}

// NewClientCredentialReader builds the read model over a client.Reader.
func NewClientCredentialReader(r client.Reader) *ClientCredentialReader {
	return &ClientCredentialReader{reader: r, teamNS: map[string]string{}}
}

// Credentials resolves the Team by UID (same rename-proof scoping as overview.go), then projects
// every Agent in the Team's namespace into a credential row, joining Paused Runs whose
// spec.agents select that Agent. Deterministic: rows sort by agent name, paused runs by name.
func (r *ClientCredentialReader) Credentials(ctx context.Context, teamUID string) (CredentialsOverview, error) {
	ns, err := r.teamNamespace(ctx, teamUID)
	if err != nil {
		return CredentialsOverview{}, err
	}

	var agents ksquadv1.AgentList
	if err := r.reader.List(ctx, &agents, client.InNamespace(ns)); err != nil {
		return CredentialsOverview{}, err
	}
	var runs ksquadv1.RunList
	if err := r.reader.List(ctx, &runs, client.InNamespace(ns)); err != nil {
		return CredentialsOverview{}, err
	}

	// Index paused credential holds by the Agents the paused Run selects. An explicit foreign
	// namespace in the ref is skipped: ObjectRef.Namespace empty means the Run's own namespace,
	// and a cross-namespace name would otherwise attribute the hold to a same-named Agent in
	// THIS namespace — wrong row, wrong banner (PR #87 review).
	holdsByAgent := map[string][]PausedRunRef{}
	for i := range runs.Items {
		run := &runs.Items[i]
		hold, ok := credentialHold(run)
		if !ok {
			continue
		}
		for _, ref := range run.Spec.Agents {
			if ref.Namespace != "" && ref.Namespace != ns {
				continue
			}
			holdsByAgent[ref.Name] = append(holdsByAgent[ref.Name], hold)
		}
	}

	out := CredentialsOverview{Team: ns}
	for i := range agents.Items {
		a := &agents.Items[i]
		row := AgentCredentialRow{
			Agent:         a.Name,
			Namespace:     a.Namespace,
			Runtime:       a.Spec.RuntimeRef.Name,
			Model:         a.Spec.Model,
			CredentialRef: a.Namespace + "/" + a.Spec.CredentialSecretRef.Name,
			// HONESTY RULE: zero credential knowledge ⇒ unknown, never a fabricated green
			// badge. Only a credential hold DOWNGRADES (to expired) today; *connected* and
			// *refreshing* are positive states the 7.7 credential controller derives once it
			// exists (ISI-2899) — the projection does not invent them.
			Health: CredHealthUnknown,
		}
		// Dual-read (ISI-3681 E3-S3 AC6, R-CR1 C2): the authoritative write target
		// is spec.credentialClass (persisted by the compose surface); fall back to
		// the ksquad.io/credential-class annotation only for legacy rows written
		// before the field round-tripped, so the Credentials screen stays honest.
		if a.Spec.CredentialClass != "" {
			row.CredentialClass = a.Spec.CredentialClass
		} else if cls, ok := a.Annotations["ksquad.io/credential-class"]; ok {
			row.CredentialClass = cls
		}
		if holds, held := holdsByAgent[a.Name]; held {
			sort.Slice(holds, func(x, y int) bool { return holds[x].Name < holds[y].Name })
			row.PausedRuns = holds
			row.Health = CredHealthExpired
		}
		out.Agents = append(out.Agents, row)
	}
	sort.Slice(out.Agents, func(x, y int) bool { return out.Agents[x].Agent < out.Agents[y].Agent })
	return out, nil
}

// teamNamespace resolves the Team UID to its namespace (the §12.1 tenancy root), memoized:
// a Team's UID is immutable for the object's lifetime, so the cluster-wide list runs at most
// once per distinct UID per reader — repeat page loads never re-deep-copy every Team
// (PR #87 review perf note; overview.go keeps its inline shape until it adopts the same memo).
func (r *ClientCredentialReader) teamNamespace(ctx context.Context, teamUID string) (string, error) {
	if teamUID == "" {
		return "", ErrTeamNotFound
	}
	r.mu.RLock()
	ns, hit := r.teamNS[teamUID]
	r.mu.RUnlock()
	if hit {
		return ns, nil
	}
	// Write lock spans resolve+store so a cold-start burst runs the cluster-wide
	// List at most once; the re-check keeps repeat callers on the memoized path.
	r.mu.Lock()
	defer r.mu.Unlock()
	if ns, ok := r.teamNS[teamUID]; ok {
		return ns, nil
	}
	var teams ksquadv1.TeamList
	if err := r.reader.List(ctx, &teams); err != nil {
		return "", err
	}
	for i := range teams.Items {
		if string(teams.Items[i].UID) == teamUID {
			ns := teams.Items[i].Namespace
			r.teamNS[teamUID] = ns
			return ns, nil
		}
	}
	return "", ErrTeamNotFound
}

// credentialHold extracts the paused-on-credential signal from one Run: phase Paused AND a
// Paused-type condition whose reason names a credential failure. Anything else (rate_limited,
// pending, running) is not a credential hold.
func credentialHold(run *ksquadv1.Run) (PausedRunRef, bool) {
	if run.Status.Phase != ksquadv1.RunPhasePaused {
		return PausedRunRef{}, false
	}
	for i := range run.Status.Conditions {
		c := &run.Status.Conditions[i]
		if c.Type != "Paused" || c.Status != metav1.ConditionTrue {
			continue
		}
		if isCredentialHold(c.Reason) {
			ref := PausedRunRef{Name: run.Name, Reason: c.Reason}
			if !c.LastTransitionTime.IsZero() {
				t := c.LastTransitionTime.Time
				ref.Since = &t
			}
			return ref, true
		}
	}
	return PausedRunRef{}, false
}

// credentials is the handler behind GET /api/credentials. Rides the §13 BFF authz choke point
// (mounted in routes); the projection is scoped to the caller's Team and reads nothing from the
// request beyond the stamped AuthorContext.
func (s *Server) credentials(reader CredentialOverviewReader) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		auth, ok := discussion.AuthFromContext(r.Context())
		if !ok || auth.Principal == "" {
			writeJSONError(w, http.StatusUnauthorized, "unauthenticated")
			return
		}
		overview, err := reader.Credentials(r.Context(), auth.TeamID.String())
		if errors.Is(err, ErrTeamNotFound) {
			writeJSONError(w, http.StatusNotFound, "no credential surface for this team")
			return
		}
		if err != nil {
			writeJSONError(w, http.StatusBadGateway, "credential read model unavailable")
			return
		}
		writeJSON(w, http.StatusOK, overview)
	}
}

// connectClaude is the handler behind POST /api/credentials/connect — the 7.7 one-click
// browser-OAuth seam. The OAuth flow itself (login URL minting, callback, Secret write-back) is
// ISI-2899's build; until it lands this answers the DOCUMENTED 501 (house pattern: the route is
// part of the contract, only its backing is pending) so the console button has one honest
// endpoint to call instead of a fabricated flow.
func (s *Server) connectClaude() http.HandlerFunc {
	return func(w http.ResponseWriter, _ *http.Request) {
		writeJSON(w, http.StatusNotImplemented, map[string]string{
			"error":    "not implemented",
			"detail":   "Connect Claude (zero-touch OAuth lifecycle, story 7.7) is not yet hosted by the apiserver",
			"tracking": "ISI-2899: credential controller + Connect Claude OAuth flow",
		})
	}
}
